marlowe-runtime/docs/filtered-chain-sync-spec-1.0.md

```diff
@@ -0,0 +1,375 @@
+implementation that satisfies all the MUST or REQUIRED level and all the SHOULD
+level requirements for its protocols is said to be "unconditionally compliant";
+one that satisfies all the MUST level requirements but not all the SHOULD level
+requirements for its protocols is said to be "conditionally compliant."
+
+### Terminology
+
+The following terminology is used to refer to the roles of participands in, and
+objects of, the protocol.
+
+- connection: A transport layer channel over which a client and a server exchange messages.
+- client: The role played by the participant which sends queries and receives chain data.
+- server: The role played by the participant which responds to queries maintains chain data.
+- state: A member of the set of states specified by the state machine of the protocol.
+- message: A member of the set of state transitions specified by the state maching of the protocol.
+- payload: Data associated and sent with a message.
+- query: Part of the payload of a query next message which the client sends to the server to tell it what to send next.
+- error: A protocol level error which is triggered when one party sends an invalid message. Results in the termination of the connection.
+- rejection: An application level error which is triggered when the server rejects a syntactically valid, semantically invalid query sent by the client. Does not result in the termination of the connection.
+- roll forward: A message sent by the server to the client in response to a query.
+- result: Part of the payload of a roll forward message which contains the result of a query.
+- roll backward: A messagge sent by the server to the client when the client's position has been rolled back.
+- tip: The latest chronological point in the blockchain
+- point: A point in the blockchain up to and including the tip
+- position: The point where a client is currently "focused."
+- schema version: A named and numberedversion of the query schema supported by the parties of the protocol. Used in the handshake to verify that the server supports the query schema supported by the client.
+- handshake: A proceedure used to establish a connection which ensures the client and server understand the same queeries.
+- agency: Used to describe which role is permited to send a message in a given state.
+
+### Operation
+
+The Filtered Chain Sync Protocol is a stateful client-server protocol. When a
+client establishes a connection with a server, a handshake is performed. If the
+handshake succeeds, the idle state is reached. From the idle state, the client
+has the agency to send a query or terminate the connection. If the client
+terminates the connection, by sending a `Done` message, then the server closes
+the connection and the protocol terminates. If the client sends a query, by
+sending a `QueryNext` message, the server can choose to respond immediately to
+the query, reject the query with a rejection reason, or ask the client to wait
```

Comment:
Do we really need `Wait`? It just tells the client to do wait it would be doing anyway. Does it have information content? Surely the underlying transport layer would ensure message receipt, keep-alives/pings, etc.

Reply:
This is from the original Chain Sync protocol. The server sends a `Wait` message to the client when it reaches the tip, which lets the client know it may take an arbitrary amount of time to receive a response. I've always thought this was a somewhat odd design decision myself, but I decided to preserve it.

Reply:
I think the idea is to give the client a chance to run an action when it is informed it must wait for a reply.

Reply:
Thanks for reminding me. Yes, let's keep this, both for consistency and for signaling the client that they might experience a long wait.